Upgrade helper

sgallagher opened the pull request:
Add Modulemd.UpgradeHelper object
    
This is the implementation of an object that will hold known
module:stream values to provide a way to resolve ambiguous
upgrades from ModuleStreamV2 dependencies.

Signed-off-by: Stephen Gallagher <sgallagh@redhat.com>

sgallagher commented:
This is the basic object; I still need to work on associating it with the `ModuleIndex` and `Module` objects. But if you could review the object in the meantime, that would be helpful.

mmathesius commented on modulemd/include/modulemd-2.0/modulemd-upgrade-helper.h:
```diff
@@ -0,0 +1,104 @@
+ * for.
+ *
+ * Returns: (transfer full): A list of known streams to provide clues to the
+ * stream upgrade process.
```
```suggestion
 * stream upgrade process. NULL if there are no known streams for @module_name.
```
It's probably worth adding a quick test that querying an unknown module returns NULL, too.

sgallagher replied:
I went the other way; it turns out that PyGObject always returns `[ ]` there, even if it gets NULL back. I decided to keep it consistent and I'll return a zero-length array to C.

mmathesius reviewed (commented):
Just one nit-pick on documentation and a suggested test to add.

sgallagher commented:
I fixed up the requested issue and added the test.

I also included two other patches, another one for fixing up Travis tests (since this PR showed it was still not picking up the overrides properly) and also clean up some memory issues detected by clang.

mmathesius commented on modulemd/modulemd-yaml-util.c:
```diff
@@ -504,6 +504,17 @@ modulemd_yaml_parse_string_set (yaml_parser_t *parser, GError **error)
     }
 
+  /* Work around false-positive in clang static analysis which thinks it's
+	   * possible for this function to return NULL and not set error.
+	   */
```
Odd comment indentation...

sgallagher replied:
Fixed

mmathesius reviewed (approved):
Aside from the noted odd comment indentation, this looks good to me. Thanks!

mmathesius commented on modulemd/modulemd-yaml-util.c:
```diff
@@ -504,6 +504,17 @@ modulemd_yaml_parse_string_set (yaml_parser_t *parser, GError **error)
     }
 
+  /* Work around false-positive in clang static analysis which thinks it's
+	 * possible for this function to return NULL and not set error.
+	 */
```
Indentation is still weird--caused by a hard tab at the beginning of those two lines.

sgallagher replied:
Thanks, it looked fine on my editor because the tabs were four spaces wide. Fixed now.

mmathesius replied:
No problem. I thought `clang-format` expanded hard tabs...

sgallagher replied:
It doesn't touch tabs that are inside a comment block.
